### DISCUSSION TITLE
Sharing some keybindings/functions I've made

### DISCUSSION THREAD

**Discussion by lmburns** [sayanarijit/xplr#241] [Show and tell]
These are some of the ideas I have came up with over the past couple days. I come from using [`lf`](https://github.com/gokcehan/lf) (which I have an extension configuration) and am trying to implement as many features as I can from there. If anyone is interested in my `lf` configuration, it can be found [here](https://github.com/lmburns/dotfiles/blob/master/.config/lf/lfrc). Also, the full configuration for `xplr` can be found [here](https://github.com/lmburns/dotfiles/blob/master/.config/xplr/init.lua).

I may update this post whenever I add more.

Firstly, using `H` and `L` to go back and forth in the path.
```lua
key = xplr.config.modes.builtin.default.key_bindings.on_key
key.L = key["ctrl-i"] -- next visited path
key.H = key["ctrl-o"] -- prev visited path
```

Jump to any directory using `zoxide`. `autojump` can be used as well, by replacing the `zoxide` command with `'PTH=$(autojump -s | tac | awk 'NR>7{print $2}' | fzf)'`.
```lua
key.n = {
  help = "zoxide jump",
  messages = {
    { BashExec = [===[
      PTH=$(zoxide query -i)
      if [ "$PTH" ]; then
        echo ChangeDirectory: "'"${PTH:?}"'" >> "${XPLR_PIPE_MSG_IN:?}"
      fi
    ]===] }
  }
}
```

Creating a new session that starts with `iTerm2`:
```lua
-- New Session: iTerm2
key["ctrl-n"] = {
  help = "new session",
  messages = {
    { BashExecSilently = [[
      osascript <<EOF
      tell application "iTerm2"
        tell current window
          create tab with default profile
          tell current session to write text "xplr"
        end tell
      end tell
    ]] },
    "Refresh",
  },
}
```

A `bookmark` mode that allows for a bookmark file to be used throughout multiple sessions. It is set to the environment variable `XPLR_BOOKMARK_FILE`. A bookmark can be added, deleted, or jumped to.
```lua
-- Bookmark: mode binding
cmodes["bookmark"] = {
  name = "bookmark",
  key_bindings = {
    on_key = {
      m = {
        help = "bookmark dir",
        messages = {
          { BashExecSilently = [[
          PTH="${XPLR_FOCUS_PATH:?}"
          if [ -d "${PTH}" ]; then
            PTH="${PTH}"
          elif [ -f "${PTH}" ]; then
            PTH="$(dirname "${PTH}")"
          fi
          if echo "${PTH:?}" >> "${XPLR_BOOKMARK_FILE:?}"; then
            echo "LogSuccess: ${PTH:?} added to bookmarks" >> "${XPLR_PIPE_MSG_IN:?}"
          else
            echo "LogError: Failed to bookmark ${PTH:?}" >> "${XPLR_PIPE_MSG_IN:?}"
          fi
        ]] },
        },
      },
      g = {
        help = "go to bookmark",
        messages = {
          {
            BashExec = [===[
            PTH=$(cat "${XPLR_BOOKMARK_FILE:?}" | fzf --no-sort)
            if [ "$PTH" ]; then
              echo FocusPath: "'"${PTH:?}"'" >> "${XPLR_PIPE_MSG_IN:?}"
            fi
            ]===]
          },
        },
      },
      d = {
        help = "delete bookmark",
        messages = {
          { BashExec = [[
          PTH=$(cat "${XPLR_BOOKMARK_FILE:?}" | fzf --no-sort)
          sd "$PTH\n" "" "${XPLR_BOOKMARK_FILE:?}"
        ]] },
        },
      },
      esc = {
        help = "cancel",
        messages = {
          "PopMode",
        },
      },
    },
  },
}
```

Another bookmark manager type thing, taken from [wfxr's zsh plugin](https://github.com/wfxr/formarks) which has colored output with `fzf`. I use this as well as the other bookmark manager above because I already use this with `zsh` as it is.
```lua
gokey = xplr.config.modes.builtin.go_to.key_bindings.on_key
-- Jump: formarks (taken from wfxr zsh plugin)
gokey.b = {
  help = "bookmark jump",
  messages = {
    { BashExec = [===[
    field='\(\S\+\s*\)'
    esc=$(printf '\033')
    N="${esc}[0m"
    R="${esc}[31m"
    G="${esc}[32m"
    Y="${esc}[33m"
    B="${esc}[34m"
    pattern="s#^${field}${field}${field}${field}#$Y\1$R\2$N\3$B\4$N#"
    PTH=$(sed 's#: # -> #' "$PATHMARKS_FILE"| nl| column -t \
    | gsed "${pattern}" \
    | fzf --ansi \
        --height '40%' \
        --preview="echo {}|sed 's#.*->  ##'| xargs $list --color=always" \
        --preview-window="right:50%" \
    | sed 's#.*->  ##')
    if [ "$PTH" ]; then
      echo ChangeDirectory: "'"${PTH:?}"'" >> "${XPLR_PIPE_MSG_IN:?}"
    fi
    ]===] }
  }
}
```

Delete and restore files with [`trash-cli`](https://github.com/andreafrancia/trash-cli). There are also notifications with [`terminal-notifier`](https://github.com/julienXX/terminal-notifier) in both steps. The `ftr` script that is referenced in the restore process is something that I wrote to integrate it with `fzf` and is also below.
```lua
delkey = xplr.config.modes.builtin.delete.key_bindings.on_key

-- Trash: delete
delkey.d = {
  help = "delete",
  messages = {
      {
        BashExec = [===[
        while IFS= read -r line; do trash-put -- "${line:?}";
        done < "${XPLR_PIPE_RESULT_OUT:?}"
        echo ExplorePwdAsync >> "${XPLR_PIPE_MSG_IN:?}"
        terminal-notifier -title "🗑️  File(s) deleted" -message "File(s) moved to trash"
        ]===],
      },
      "PopMode",
  },
}

-- Trash: restore
delkey.r = {
  help = "restore delete",
  messages = {
    {
      BashExec = [===[
      ftr; echo ExplorePwdAsync >> "${XPLR_PIPE_MSG_IN:?}"
      ]===],
    },
    "PopMode",
  },
}
```
`ftr` script:
```zsh
#!/usr/bin/env zsh

# desc: fzf trash-restore

zparseopts -D -E -a opt 'l' '-local'

if [[ "${(M)opt[@]:#-(l|-local)}" == -(l|-local) ]]; then
  file="$(trash-list \
    | rg --color=never "$(pwd)" \
    | fzf --delimiter / --with-nth 1,-1 \
    | awk '{print $3}' \
  )"
else
  file="$(trash-list \
    | fzf --delimiter / --with-nth 1,-1 \
    | awk '{print $3}' \
  )"
fi

[ -n "$file" ] && yes 0 | trash-restore "$file" || exit 1
terminal-notifier -title "Trash CLI" -message "File restored"
```

Also, I have taken the icons from `lf` and put them with `xplr`
<details>
<summary><b>Icons</b></summary>
```lua
    xplr.config.node_types.special = {
      ["Cargo.toml"] = {
        meta = { icon = "⚙" },
      },
      [".git"] = {
        meta = { icon = "" },
      },
      node_modules = {
        meta = { icon = "" },
      },
      ["docker-compose.yml"] = {
        meta = {
          icon = "",
        }
      }
    }

  xplr.config.node_types.mime_essence = {
    ["text"] = {
      ["plain"] = {
        meta = { icon = "" },
      }
    },
    ["video"] = {
      ["*"] = {
        meta = { icon = "" },
      }
    },
    ["image"] = {
      ["*"] = {
        meta = { icon = "" },
      },
    }
  }

  xplr.config.node_types.extension = {
    lock = {
      meta = { icon = "" },
    },
    md = {
      meta = { icon = "" },
    },
    toml = {
      meta = { icon = "" },
    },
    rs = {
      meta = { icon = "" },
    },
    elm = {
      meta = { icon = "" },
    },
    lua = {
      meta = { icon = "" },
    },
    nix = {
      meta = { icon = "" },
    },
    js = {
      meta = { icon = "" },
    },
    py = {
      meta = { icon = "" },
    },
    txt = {
      meta = { icon = "" },
    },
    csv = {
      meta = { icon = "" },
    },
    xlsx = {
      meta = { icon = "" },
    },
    odf = {
      meta = { icon = "" },
    },
    yml = {
      meta = { icon = "⚙" },
    },
    yaml = {
      meta = { icon = "⚙" },
    },
    c = {
      meta = { icon = "" },
    },
    cc = {
      meta = { icon = "" },
    },
    clj = {
      meta = { icon = "" },
    },
    ccp = {
      meta = { icon = "" },
    },
    coffee = {
      meta = { icon = "" },
    },
    cpp = {
      meta = { icon = "" },
    },
    css = {
      meta = { icon = "" },
    },
    d = {
      meta = { icon = "" },
    },
    dart = {
      meta = { icon = "" },
    },
    erl = {
      meta = { icon = "" },
    },
    exs = {
      meta = { icon = "" },
    },
    fs = {
      meta = { icon = "" },
    },
    go = {
      meta = { icon = "" },
    },
    h = {
      meta = { icon = "" },
    },
    hh = {
      meta = { icon = "" },
    },
    hpp = {
      meta = { icon = "" },
    },
    hs = {
      meta = { icon = "" },
    },
    html = {
      meta = { icon = "" },
    },
    h = {
      meta = { icon = "" },
    },
    java = {
      meta = { icon = "♨" },
    },
    kt = {
      meta = { icon = "🅺" },
    },
    jl = {
      meta = { icon = "" },
    },
    js = {
      meta = { icon = "" },
    },
    json = {
      meta = { icon = "" },
    },
    php = {
      meta = { icon = "" },
    },
    pl = {
      meta = { icon = "" },
    },
    pro = {
      meta = { icon = "" },
    },
    rb = {
      meta = { icon = "" },
    },
    pro = {
      meta = { icon = "" },
    },
    scala = {
      meta = { icon = "" },
    },
    ts = {
      meta = { icon = "" },
    },
    vim = {
      meta = { icon = "" },
    },
    cmd = {
      meta = { icon = "" },
    },
    ps1 = {
      meta = { icon = "" },
    },
    sh = {
      meta = { icon = "" },
    },
    zsh = {
      meta = { icon = "" },
    },
    bash = {
      meta = { icon = "" },
    },
    fish = {
      meta = { icon = "" },
    },
    tar = {
      meta = { icon = "" },
    },
    tgz = {
      meta = { icon = "" },
    },
    arc = {
      meta = { icon = "" },
    },
    arj = {
      meta = { icon = "" },
    },
    taz = {
      meta = { icon = "" },
    },
    lha = {
      meta = { icon = "" },
    },
    lz4 = {
      meta = { icon = "" },
    },
    lzh = {
      meta = { icon = "" },
    },
    tlz = {
      meta = { icon = "" },
    },
    txz = {
      meta = { icon = "" },
    },
    tzo = {
      meta = { icon = "" },
    },
    t7z = {
      meta = { icon = "" },
    },
    lzma = {
      meta = { icon = "" },
    },
    zip = {
      meta = { icon = "" },
    },
    z = {
      meta = { icon = "" },
    },
    dz = {
      meta = { icon = "" },
    },
    gz = {
      meta = { icon = "" },
    },
    lrz = {
      meta = { icon = "" },
    },
    lz = {
      meta = { icon = "" },
    },
    lzo = {
      meta = { icon = "" },
    },
    xz = {
      meta = { icon = "" },
    },
    zst = {
      meta = { icon = "" },
    },
    tzst = {
      meta = { icon = "" },
    },
    bz2 = {
      meta = { icon = "" },
    },
    bz = {
      meta = { icon = "" },
    },
    tbz = {
      meta = { icon = "" },
    },
    tbz2 = {
      meta = { icon = "" },
    },
    tz = {
      meta = { icon = "" },
    },
    deb = {
      meta = { icon = "" },
    },
    rpm = {
      meta = { icon = "" },
    },
    jar = {
      meta = { icon = "" },
    },
    war = {
      meta = { icon = "" },
    },
    ear = {
      meta = { icon = "" },
    },
    sar = {
      meta = { icon = "" },
    },
    rar = {
      meta = { icon = "" },
    },
    alz = {
      meta = { icon = "" },
    },
    ace = {
      meta = { icon = "" },
    },
    zoo = {
      meta = { icon = "" },
    },
    cpio = {
      meta = { icon = "" },
    },
    z7z = {
      meta = { icon = "" }, -- can't start with number?
    },
    rz = {
      meta = { icon = "" },
    },
    cab = {
      meta = { icon = "" },
    },
    tbz2 = {
      meta = { icon = "" },
    },
    wim = {
      meta = { icon = "" },
    },
    swm = {
      meta = { icon = "" },
    },
    dwm = {
      meta = { icon = "" },
    },
    esd = {
      meta = { icon = "" },
    },
    pdf = {
      meta = { icon = "" },
    },
  }
```
</details>

**Comment by sayanarijit** [comment]
Nice! Thanks for sharing...

> xplr doesn't return anything on close unless pressing # 

By default, `enter` is mapped to `PrintResultAndQuit`.

> -- can't start with number?

Use the ["7z"] syntax.

I had a quick glance over your config. 2 things can be improved:

- `XPLR_PIPE_FOCUS_OUT` has been deprecated in favor of `XPLR_FOCUS_PATH` (sorry the pipe doc was outdated)
- With the latest version, you don't need to send `Refresh` manually anymore.

**Reply by lmburns** [reply]
Okay, awesome! I got that to work. Is there a variable/pipe that contains the `cwd` that is able to print to `stdout`? I found `PrintAppStateAndQuit` and `PrintResultAndQuit`. I also tried something like
```bash
{ BashExec = [[ cat "${XPLR_PIPE_DIRECTORY_NODES_OUT:?}" ]] }
```
I would use the `q` keybinding for that if there was. However, I did write a function similar to `lf`'s `cd` on exit, which has to be placed in my `.zshrc` for it to work, and it is somewhat similar:
```zsh
xd() {
  pth="$(xplr)"
  if [[ "$pth" != "$PWD" ]]; then
    if [[ -d "$pth" ]]; then
      cd "$pth"
    elif [[ -f "$pth" ]]; then
      cd "$(dirname "$pth")"
    fi
  fi
}
```
Also, thank you for the tips! I'm changing them now

**Reply by sayanarijit** [reply]
When in the shell (`:!`), run `env | grep XPLR` to see the complete list of exported variables. Btw in xplr, pwd and cwd are synchronised. Please create an issue if you want them to be separated, explaining the use case.

**Reply by sayanarijit** [reply]
Because it's synchronised,  you can use `echo $PWD` and then `Quit`.

**Reply by lmburns** [reply]
Okay, I updated my comment and updated my bookmark command above to add the actual directory to bookmark file instead of the full path.

And I guess I could put in an issue for something like a `PrintDirectoryAndQuit`. The use would be say that I am am using the above `xd` function and am in a directory that I would like to `cd` into, I would either have to go up a directory and hit `enter` to `cd` into it, or I would have to find a regular file and hit `enter` on it.  That is, unless there is a way to use `BashExec` to print `$PWD` to the `tty` that started `xplr`.

Also, I would prefer to use `zsh` as the shell within `xplr`, and whenever I run a command, this error message is printed: `touch: setting times of '/dev/stderr': Permission denied`. I think this may be a bug with one of my `.zshrc` scripts or something that is being sourced, as it doesn't happen with `bash`.

```lua
key['!'] = {
  help = "shell",
  messages = {
    { Call = { command = "zsh", args = { "-i" } } },
    -- same result, with and without the 'args = {"-i"}'
    "ExplorePwdAsync",
    "PopMode",
  }
}
```

**Reply by sayanarijit** [reply]
Sure please create the issue. I'm not aure about the shell though. Seems like some zsh plugin is trying to run `touch /dev/stderr` on every command?

**Reply by max-baz** [reply]
> Because it's synchronised, you can use echo $PWD and then Quit.

When you said that we can use `echo $PWD`, could you explain please, from inside xplr can we print anything to the stdout? I saw #257 but tried something naive (below) while it's not implemented, but it doesnt print anything. Should I be echoing to some pipe?

```
    key.Q = {
        help = "quit cd",
        messages = {
            { BashExecSilently = "echo $PWD" },
            "Quit",
        },
    }
```

And by the way, nice idea with `PrintPwdAndQuit`, I would personally bind it to `Q` and then combine this with the ideas from other threads to have my final `alias x='cd "${$(PAGER="less -+F" xplr):-.}"'`, with which I could do something I could not do in nnn: decide whether I want to "just quit" or "quit and cd" AFTER I launch explorer (with pressing `q` or `Q`).

**Reply by sayanarijit** [reply]
>  could you explain please, from inside xplr can we print anything to the stdout?

My bad... It's not possible to do this directly. `PrintPwdAndQuit` seems to be the way. But I'm still worried there are already 3 key maps for quitting. Adding more will get really confusing.

I also have another solution in mind, i.e. adding CLI argument like `--pwd-as-result`. This way we can add as many quit options as we want, without expanding the key bindings.

**Reply by max-baz** [reply]
as long as it's possible to configure them, I don't mind if not all keys will be by default 😉 For me personally it will be a huge benefit if I can decide how I want to quit not *before* I launch xplr but *after*, but that means I must prepare all the shortcuts for myself 😉

**Reply by sayanarijit** [reply]
Interesting... I guess we can implement both. "Choose what you prefer" :smile:

**Reply by max-baz** [reply]
inability to choose whether I want to "just quit" or "quit cd" in runtime (after launch) was always one of my big annoyances with nnn, btw 🙂

**Reply by sayanarijit** [reply]
Hey, I am converting the hacks to plugins e.g. https://github.com/sayanarijit/zoxide.xplr, and listing them [here](https://arijitbasu.in/xplr/en/awesome-plugins.html). Let me know if you create any or want to maintain any of the plugins I created.

**Reply by sayanarijit** [reply]
Also check out https://github.com/sayanarijit/xplr/pull/437, probably the best "cd on quit" functionality.

**Comment by sayanarijit** [comment]
I have added them in the [Hacks Wiki](https://github.com/sayanarijit/xplr/wiki/Hacks). Someday, they'll probably become independent portable plugins.

**Reply by lmburns** [reply]
Awesome! I can update this thread whenever I add more. I just realized once I read over those plugins in the wiki that on the wfxr bookmark plugin that the line with `xargs $list` actually needs to be `xargs exa` instead. So, requires [`exa`](https://github.com/ogham/exa). Also, that one doesn't require `sd`, it's the bookmark one above it that I have used `sd`. It can be changed as well, I just used it for its convenient syntax.

Then the `iTerm` tab one says it requires `zoxide` and it doesn't require anything there except to be on macOS.

Lastly, just pointing out a spelling mistake I noticed near the top on the 4th line: `fzf integration tutotial`

**Reply by sayanarijit** [reply]
Thanks. Fixed them all (I hope).

**Comment by sayanarijit** [comment]
Made a little correction in the [trash-cli hack](https://github.com/sayanarijit/xplr/wiki/Hacks#trash-trash-cli). Using `BashExecSilently` instead of `BashExec` where interaction isn't required, helps avoid the little flickering of screen.

**Comment by lmburns** [comment]
I changed it in my configuration and yeah I see the difference. Is it possible to make `xplr` use `LSCOLORS` environment variable to look something like this
<img width="274" alt="Screen Shot 2021-06-09 at 11 33 58 PM" src="https://user-images.githubusercontent.com/44355502/121465363-36bbdb80-c9a5-11eb-8ffa-8aaee66b372c.png">

**Reply by sayanarijit** [reply]
xplr doesn't read LSCOLORS. But you can do this with lua. See the default lua functions. It's documented in the UI config.

**Comment by sayanarijit** [comment]
Hey @lmburns, I just converted the Alacritty integration hack to this [alacritty.xplr](https://github.com/sayanarijit/alacritty.xplr) plugin with some added features. If you want, you can apply the concept to iTerm2.
See [writing plugins](https://arijitbasu.in/xplr/en/writing-plugins.html).

**Reply by sayanarijit** [reply]
Also, since I'm not using zoxide a lot, if you want, you can maintain [zoxide.xplr](https://github.com/sayanarijit/zoxide.xplr) i.e. I can transfer the repo.

**Reply by lmburns** [reply]
That would be fine with me. I've never done a transferring of repos, so I'm not sure how to

**Reply by sayanarijit** [reply]
Sent req.
